Fix test errors related with authorization

elliotyim opened the pull request:
JWT를 적용하고 난 뒤 모든 API 접근에 권한이 필요해져서
테스트 코드에서 401 에러가 나는 부분을 우선적으로 수정했습니다.
(JWT 토큰을 헤더에 넣어서 요청하도록 변경)

추가: HTTPS로 돌아가는 서버 환경에서도 테스트코드에 에러가 나지 않도록 secure=True 옵션 추가했습니다.

elliotyim committed:
Fix test errors related with authorization

elliotyim committed:
Make test code accept ssl request

elliotyim committed:
Apply jwt token to test codes

elliotyim committed:
Give permissions to APIs

elliotyim commented:
테스트 코드에 옵션 두 개가 추가되었습니다.
- HTTP_AUTHORIZATION=token: JWT 토큰을 헤더에 넣는 옵션
- secure=True: 도메인이 HTTPS일 때 테스트 코드가 통과하도록 만드는 옵션

그리고, User에서 Authentication에 JWT가 추가되었기에 Post API가 GET 방식을 제외하고
로그인된 사용자만 API에 접근할 수 있도록 변경했습니다.

elliotyim commented on post/api/post.py:
```diff
@@ -16,6 +17,19 @@
+    def get_permissions(self) -> List[BasePermission]:
+        permission_classes: List[Union[Type[AllowAny], Type[IsAuthenticated]]] = list()
+
+        if self.action in ['list', 'retrieve']:
```
여기서 self.action이 mypy 때문에 현재 클래스에 action이라는 변수가 없다고 나와서
NamedTuple을 상속하려고 했는데 ViewSet을 이미 상속받은 상태라 사용할 수 없다고 나옵니다.
그래서 생성자 쪽에 무의미한 3줄이 추가되었는데, 뭔가 좋은 방법이 없을까요?

hyeonjames replied:
의도를 알기가 힘든거 같은데, NamedTuple 안써도 기본값은 넣을 수 있어요..
```python
class PostAPI(viewsets.ViewSet):
     action: Optional[str] = None
```
생성자로 action을 설정해주고 싶은 거라면, 그냥 생성자를 넣는게 좋을 것 같아요.

elliotyim replied:
> 의도를 알기가 힘든거 같은데, NamedTuple 안써도 기본값은 넣을 수 있어요..
> 
> ```python
> class PostAPI(viewsets.ViewSet):
>      action: Optional[str] = None
> ```
> 
> 생성자로 action을 설정해주고 싶은 거라면, 그냥 생성자를 넣는게 좋을 것 같아요.

헐...그러네요. 난 왜 저렇게 쓴거지?;;
감사합니다!!

elliotyim reviewed (commented):
도움!

elliotyim committed:
Polish codes

hyeonjames commented on tests/post/jwt_token.py:
```diff
@@ -0,0 +1,12 @@
+from user.models import User
+
+
+def get_jwt_token_of(user: User) -> str:
```
모듈을 설계할 때 위상도 꽤나 중요한데요, 이 모듈은 jwt_token에 관한건데 꼭 post 위치에 있을 이유가 없을 것 같아요.

elliotyim replied:
> 모듈을 설계할 때 위상도 꽤나 중요한데요, 이 모듈은 jwt_token에 관한건데 꼭 post 위치에 있을 이유가 없을 것 같아요.

아 그렇네요 ㄷㄷ;;
다른 API에서 쓸 수도 있으니

hyeonjames commented on tests/post/test_comment.py:
```diff
@@ -1,28 +1,39 @@
     pytestmark = pytest.mark.django_db
 
     def test_create_comment(self, api_client, users, posts):
+        token = f'JWT {get_jwt_token_of(users[0])}'
```
token을 왠만하면 fixture로 구현하는게 어떨까요?

elliotyim replied:
> token을 왠만하면 fixture로 구현하는게 어떨까요?

fixture로 만든다는게 같은 파일 내에서 token을 가져올 수 있게 하는건가요?

토큰이 특정 유저의 정보 (ID, Password)를 가지고 만드는거라서,
만약 글을 쓴 유저만 해당 글을 지울 수 있게 하려면 토큰이 해당 유저가 발급한 것이여야만 해서요.
test 내에서 그 부분을 명시할 필요가 있지 않을까해서 이런 방식으로 만들었습니다.

제가 아직 fixture쪽에 대한 이해도가 부족해서 ㄷㄷ..
fixture로 만들어서 주입받는다면 어떤식으로 하면 될까요?

hyeonjames replied:
어짜피 test 쪽은 어느 유저가 되도 상관이 없을거 같은데, 특정 유저로 제한하는게 아니라면 fixture 로 만들어도 상관이 없을 거에요. 그냥 users fixture 랑 같이 쓰면 되지 않나요?

```python
@pytest.fixture
def token(users):
   return f'JWT {get_jwt_token_of(users[0]}' 
```

elliotyim replied:
> 어짜피 test 쪽은 어느 유저가 되도 상관이 없을거 같은데, 특정 유저로 제한하는게 아니라면 fixture 로 만들어도 상관이 없을 거에요. 그냥 users fixture 랑 같이 쓰면 되지 않나요?
> 
> ```python
> @pytest.fixture
> def token(users):
>    return f'JWT {get_jwt_token_of(users[0]}' 
> ```

아 그렇네요.
이렇게 변경하겠습니다.

hyeonjames reviewed (changes requested):
코멘트 몇개 달았어요~

elliotyim committed:
Modify the way to get JWT on test

Change Log:
- Rename jwt module properly.
- Moved jwt module into `common` app.
- The JWT is given by the fixture now.

elliotyim committed:
Remove unnecessary fixture parameters

elliotyim commented:
get_jwt_of() 는 아무래도 이 PR에서만 임시로 사용되고 변경되거나 지워질 것 같습니다.
#43 에서 변경된 simplejwt는 다른 방식으로 token을 얻네요.

(참고: https://django-rest-framework-simplejwt.readthedocs.io/en/latest/creating_tokens_manually.html)

hyeonjames reviewed (approved):
수고하셨어요~
